modules/navier_stokes/src/fv/kernels/INSFVMomentumAdvection.C

```diff
@@ -0,0 +1,505 @@
+  // Note that diffusion coefficients should always be positive, e.g. elliptic operators always
+  // yield positive definite matrices
+  //
+  // Example 1D discretization of diffusion, e.g. the sum of the fluxes around a control volume:
+  //
+  // \sum_f -D \nabla \phi * \hat{n} =
+  //   -D_e * (phi_E - \phi_C) / d_{CE} * 1 - D_w * (\phi_C - \phi_W) / d_{WC} * -1 =
+  //   D_w / d_{CE} * (\phi_C - \phi_E) + D_e / d_{WC} * (\phi_C - \phi_W)
+  //
+  // Note the positive coefficients for \phi_C !!
+  //
+  // Now an example 1D discretization for advection using central differences, e.g. an average
+  // interpolation
+  //
+  // \sum_f \vec{u} \phi \hat{n} =
+  //   u_w * (\phi_W + \phi_C) / 2 * -1 + u_e * (\phi_C + \phi_E) / 2 * 1 =
+  //   -u_w / 2 * \phi_W + u_e / 2 * \phi_E + (u_e - u_w) / 2 * \phi_C
+  //
+  // Note that the coefficient for \phi_C may or may not be positive depending on the values of u_e
+  // and u_w
+
+  VectorValue<ADReal> coeff = 0;
+
+  ADRealVectorValue elem_velocity(_u_var->getElemValue(&elem));
+
+  if (_v_var)
+    elem_velocity(1) = _v_var->getElemValue(&elem);
+  if (_w_var)
+    elem_velocity(2) = _w_var->getElemValue(&elem);
+
+  auto action_functor = [&coeff, &elem_velocity, &mu, this](const Elem & /*functor_elem*/,
+                                                            const Elem * const neighbor,
+                                                            const FaceInfo * const fi,
+                                                            const Point & surface_vector,
+                                                            Real /*coord*/,
+                                                            const bool /*elem_has_info*/) {
+    mooseAssert(fi, "We need a non-null FaceInfo");
+
+    const bool rc_elem_is_fi_elem = (neighbor == fi->neighborPtr());
+    const Point normal = rc_elem_is_fi_elem ? fi->normal() : Point(-fi->normal());
+
+    // Unless specified otherwise, "elem" here refers to the element we're computing the
+    // Rhie-Chow coefficient for. "neighbor" is the element across the current FaceInfo (fi)
+    // face from the Rhie-Chow element
+
+    if (fi->isBoundary())
+    {
+      // In my mind there should only be about one bc_id per FaceInfo
+      for (const auto bc_id : fi->boundaryIDs())
+      {
+        if (_no_slip_wall_boundaries.find(bc_id) != _no_slip_wall_boundaries.end())
+        {
+          // Need to account for viscous shear stress from wall
+          for (const auto i : make_range(_dim))
+            coeff(i) += mu * surface_vector.norm() /
+                        std::abs((fi->faceCentroid() - fi->elemCentroid()) * normal) *
+                        (1 - normal(i) * normal(i));
+
+          // No flow normal to wall, so no contribution to coefficient from the advection term
+          break;
+        }
+
+        if (_flow_boundaries.find(bc_id) != _flow_boundaries.end())
```

Comment:
this case is split in inlet / outlet and then further split in V_known / V_and_P_known / P_known for inlet and P_known / mdot_known / fully_developped in Moukalled
I ll need time to look at it further

Reply:
The different treatment to me reflects an assumption of an upwind scheme in Moukalleds derivation. The implementation here is generic and should yield the correct result whether an average or upwind interpolation is chosen for the advected quantity (or some other yet to be implemented scheme)

Reply:
OK after reading it more:
- there are special provisions for inlets with the velocity direction and pressure specified. We dont have an immediate need for those so I think it's fine to pass now
- the main distinction is actually between inlets and fully-developed outlets. The viscous flux is zeroed out on fully developed outlets, and the boundary gradient value is made orthogonal to the surface.

Reply:
gradient of what?

Reply:
velocity, see 15.140

Reply:
Ok you are correct; the code I currently have here is only valid for the Moukalled cases of:

- outflow: specified static pressure (fully developed outflow assumed)
- outflow: fully developed outflow (static pressure computed via 15.145)

Reply:
I do have to say that I don't understand why 15.140 "ensures that the gradient along the boundary surface vector is zero". I only see how the math works out, e.g. `\nabla v_b \cdot \hat{n} = 0`, if `\hat{e}_b = \hat{n}`

Reply:
Is it possible that this is an error? They've given a formula that guarantees that `\nabla v_b \cdot \hat{e}_b = 0`. Perhaps `\hat{e}_b` should be replaced with `\hat{n}` ?

Reply:
\nabla v_b \cdot \hat{n} = 0 is a zero flux BC and I think it makes sense here. In the text it says zero flux BCs are similar to what they do. 
I dont think it's an error, it says to use the element value at the boundary, regardless of the shape of the element. So eb is not the same as the normal. 

I agree that using the normal instead makes a lot of sense. Maybe their treatment is easier? I'm not sure

Reply:
Which to me goes against other language in that section which says they want to zero the normal gradient but *allow* a tangential gradient, e.g. if `\hat{e_b} != \hat{n}`, then `v_b` should not necessarily be equal to `v_c`

Reply:
I really think that 15.140 is not correct as written

Reply:
I think I'm convinced. Just a note that they can still allow a tangeantial gradiant by using e_b. It's zeroing the normal that's not done.

Reply:
Aha! See equation 15.152 has the form that I would expect 15.140 to have